bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -524,32 +535,208 @@ impl StateDriver {
     async fn start_vm(
         &mut self,
         start_reason: VmStartReason,
-    ) -> anyhow::Result<()> {
+    ) -> VmStartOutcome {
         info!(self.log, "starting instance"; "reason" => ?start_reason);
 
-        let start_result =
-            self.objects.lock_exclusive().await.start(start_reason).await;
+        // Tell listeners that the VM's components are now starting up and not
+        // merely being created (but keep the VM in the Migrating state if it's
+        // being started pursuant to a migration in).
+        if let VmStartReason::ExplicitRequest = start_reason {
+            self.external_state
+                .update(ExternalStateUpdate::Instance(InstanceState::Starting));
+        }
 
-        self.input_queue.notify_request_completed(CompletedRequest::Start {
-            succeeded: start_result.is_ok(),
-        });
+        // The start sequence is arranged so that calls to block backends can be
+        // interleaved with processing of requests from the external request
+        // queue. This allows Nexus to reconfigure Crucible backends while they
+        // are being activated, which can be necessary if the VM's original
+        // specification specifies a Crucible downstairs server that is offline
+        // or unavailable. (Downstairs instances can disappear at any time,
+        // e.g. due to sled failure, so these configurations aren't necessarily
+        // client errors.)
+        //
+        // Before getting into any of that, handle the synchronous portions of
+        // VM startup. First, ensure that the kernel VM and all its associated
+        // devices are in the correct initial states.
+        let objects = self.objects.lock_shared().await;
+        match start_reason {
+            // If this VM is a migration target, migration will have properly
+            // initialized the vCPUs, but will have left the kernel VM paused.
+            // Resume it here before asking any in-kernel components to start.
+            VmStartReason::MigratedIn => objects.resume_kernel_vm(),
+
+            // If this VM is starting from scratch, its kernel VM is active, but
+            // its vCPUs have not been initialized yet.
+            VmStartReason::ExplicitRequest => objects.reset_vcpus(),
+        }
 
-        match &start_result {
-            Ok(()) => {
-                self.external_state.update(ExternalStateUpdate::Instance(
-                    InstanceState::Running,
-                ));
+        // Send synchronous start commands to all devices.
+        for (name, dev) in objects.device_map() {
+            info!(self.log, "sending start request to {}", name);
+            let res = dev.start();
+            if let Err(e) = res {
+                error!(self.log, "startup failed for {}: {:?}", name, e);
```

Comment:
nit, take it or leave it: perhaps the device name ought to be a structured field on these logs to ease in using `looker` or similar?